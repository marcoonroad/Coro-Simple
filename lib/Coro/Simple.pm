use v6;

unit module Coro::Simple;

sub coro (&block) is export {
    return sub (*@params) {
        # we explore the lazy evaluation property
        # to evaluate the computation on demand
        my @yields := (gather { block |@params }).list;
        my $index   = 0;
        # generator
        return sub ( ) {
            my $result;
            if defined @yields[ $index++ ] {
                $result = @yields[ $index - 1 ];
            }
            # otherwise, if the coroutine is dead
            else {
        	$index -= 1; # disallows further lookups
                $result = False;
            }
            return $result;
        }
    }
}

sub yield ($value) is export { take $value }

sub suspend( ) is export { take True }

# purpose:
# to check if generated value was returned from yield triggering (rather end of computation)
sub ensure (&block, $value) is export {
    if ($value ~~ Bool) && (!$value) {
        return block;
    }
    return $value;
}

# deprecated function
sub assert (&block, $value) is export {
    warn "[ Deprecated function ]\n" ~
        "This function will be removed in a future version. Please use 'ensure' instead.";
    return ensure &block, $value;
}

sub from (&generator) is export {
    my $temp = generator; # obtains the first value

    return gather {
        while ($temp !~~ Bool) || (?$temp) {
            take $temp;
            $temp = generator;
        }
    };
}

# end of module
