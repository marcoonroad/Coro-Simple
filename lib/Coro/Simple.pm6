use v6;

unit module Coro::Simple;

sub coro (&block) is export {
  my $coroutine-thread = $*THREAD;

  return sub (*@params) {
    # we explore the lazy evaluation property
    # to evaluate the computation on demand
    my @yields := (gather { block |@params }).list;
    my $index   = 0;
    # generator
    return sub ( ) {
      unless $*THREAD.id == $coroutine-thread.id {
        die "Yield called outside the owner coroutine!";
      }

      my $result;
      if defined @yields[ $index++ ] {
        $result = @yields[ $index - 1 ];
      }
      # otherwise, if the coroutine is dead
      else {
	      $index -= 1; # disallows further lookups
        $result = False;
      }
      return $result;
    }
  }
}

sub yield ($value) is export { take $value }

sub suspend( ) is export { take True }

# purpose:
# to check if generated value was returned from yield triggering (rather end of computation)
sub ensure (&block, $value) is export {
  if ($value ~~ Bool) && (!$value) {
    return block;
  }
  return $value;
}

sub from (&generator) is export {
  my $temp = generator; # obtains the first value

  return gather {
    while ($temp !~~ Bool) || (?$temp) {
      take $temp;
      $temp = generator;
    }
  };
}

# end of module
