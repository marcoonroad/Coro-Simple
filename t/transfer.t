#!/usr/bin/perl6

use v6;

# a transfer function test

use Test;
use Coro::Simple;

plan 1;

sub transfer (&generator) {
    yield &generator;
}

# impure 'begin' function
# sub begin (&generator) {
#    my $transferred = generator;
#    $transferred = $transferred( ) while $transferred;
# }

# pure 'begin' function
multi begin (Bool $) { }
multi begin (&generator) { begin generator( ) }

my $first;
my $second;
my $third;

my &ping = coro -> $msg {
    for ^3 -> $i {
	say "$msg -> $i";
	transfer $second;
    }
}

my &wtf = coro {
    for ^3 {
	say "\n" ~ "WTF?" ~ "\n\n";
	transfer $third;
    }
}

my &pong = coro -> $msg {
    for ^3 -> $i {
	say "$msg -> $i";
	transfer $first;
    }
}

$first  = ping "Ping!";
$second = wtf;
$third  = pong "Pong!";

ok $first && $second && $third;

begin $first; # begin the cycle with this generator

# a small / useful scheduler-like chunk
# (from $ping).map(&from).map: { &^generator( ) };

# for from $ping -> $coro {
#     for from $coro -> $next {
# 	$next( );
#     }
# }

# end of test
