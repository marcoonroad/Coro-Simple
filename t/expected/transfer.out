Ping! -> 0

WTF?


Pong! -> 0
Ping! -> 1

WTF?


Pong! -> 1
Ping! -> 2

WTF?


Pong! -> 2
